src/expressions/operators/compares/valueCompare.ts

```diff
@@ -326,23 +322,47 @@ function generateCompareFunction(
 
 const comparatorsByTypingKey = Object.create(null);
 
-export default function (
+function valueCompareFunction(
 	operator: string,
 	typeA: ValueType,
-	typeB: ValueType,
-	dynamicContext: DynamicContext
-): (valueA: AtomicValue, valueB: AtomicValue) => boolean {
+	typeB: ValueType
+): (valueA: AtomicValue, valueB: AtomicValue, dynamicContext: DynamicContext) => boolean {
 	// https://www.w3.org/TR/xpath-3/#doc-xpath31-ValueComp
 	const typingKey = `${typeA as number}~${typeB as number}~${operator}`;
 	let prefabComparator = comparatorsByTypingKey[typingKey];
 	if (!prefabComparator) {
 		prefabComparator = comparatorsByTypingKey[typingKey] = generateCompareFunction(
 			operator,
 			typeA,
-			typeB,
-			dynamicContext
+			typeB
 		);
 	}
 
 	return prefabComparator;
 }
+
+export function getValueCompareEvaluationFunction(
+	operator: string,
+	typeA: ValueType,
+	typeB: ValueType
+): (
+	firstSequence: ISequence,
+	secondSequence: ISequence,
+	dynamicContext: DynamicContext
+) => boolean {
+	try {
+		const evaluationFunction = valueCompareFunction(operator, typeA, typeB);
+
+		return (first, second, dynamicContext) => {
+			if (first.getLength() > 1 || second.getLength() > 1) {
+				throw new Error('XPTY0004: Sequences to compare are not singleton.');
+			}
+
+			return evaluationFunction(first.first(), second.first(), dynamicContext);
+		};
+	} catch {
```

Comment:
Not sure here but should we throw err in the catch part?

Reply:
No, this happens only during annotation and it fails if we try to compare anyatomictype and integer for example. Since this is not allowed we just return undefined so we don't use the shortcut. During actual evaluation, the error will not get caught.